ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionService.scala

```diff
@@ -0,0 +1,707 @@
+          fromAndToBegin = request.update(_.begin := ledger.begin, _.end := ledger.begin)
+          transactions <- ledger.flatTransactions(fromAndToBegin)
+        } yield {
+          assert(
+            transactions.isEmpty,
+            s"Received a non-empty stream with ${transactions.size} transactions in it.")
+        }
+    }
+
+  private[this] val serveElementsUntilCancellation =
+    LedgerTest("TXServeUntilCancellation", "Items should be served until the client cancels") {
+      ledger =>
+        val transactionsToSubmit = 14
+        val transactionsToRead = 10
+        for {
+          party <- ledger.allocateParty()
+          dummies <- Future.sequence(Vector.fill(transactionsToSubmit)(for {
+            contract <- ledger.create(party, Dummy(party))
+          } yield contract))
+          transactions <- ledger.flatTransactions(transactionsToRead, party)
+        } yield {
+          assert(
+            dummies.size == transactionsToSubmit,
+            s"$transactionsToSubmit should have been submitted but ${dummies.size} were instead")
+          assert(
+            transactions.size == transactionsToRead,
+            s"$transactionsToRead should have been received but ${transactions.size} were instead")
+        }
+    }
+
+  private[this] val deduplicateCommands =
+    LedgerTest(
+      "TXDeduplicateCommands",
+      "Commands with identical command identifier and application identifier should be accepted and deduplicated") {
+      ledger =>
+        for {
+          party <- ledger.allocateParty()
+          request <- ledger.submitAndWaitRequest(party, Dummy(party).create.command)
+          _ <- ledger.submitAndWait(request)
+          _ <- ledger.submitAndWait(request)
+          transactions <- ledger.flatTransactions(party)
+        } yield {
+          assert(
+            transactions.length == 1,
+            s"Only one transaction was expected to be seen but ${transactions.length} appeared")
+        }
+    }
+
+  private[this] val rejectEmptyFilter =
+    LedgerTest(
+      "TXRejectEmptyFilter",
+      "A query with an empty transaction filter should be rejected with an INVALID_ARGUMENT status") {
+      ledger =>
+        for {
+          party <- ledger.allocateParty()
+          request = ledger.getTransactionsRequest(Seq(party))
+          requestWithEmptyFilter = request.update(_.filter.filtersByParty := Map.empty)
+          failure <- ledger.flatTransactions(requestWithEmptyFilter).failed
+        } yield {
+          assertGrpcError(failure, Status.Code.INVALID_ARGUMENT, "filtersByParty cannot be empty")
+        }
+    }
+
+  private[this] val completeOnLedgerEnd = LedgerTest(
+    "TXCompleteOnLedgerEnd",
+    "A stream should complete as soon as the ledger end is hit") { ledger =>
+    val transactionsToSubmit = 14
+    for {
+      party <- ledger.allocateParty()
+      transactionsFuture = ledger.flatTransactions(party)
+      _ <- Future.sequence(Vector.fill(transactionsToSubmit)(for {
```

Comment:
Same remark as above regarding the for-comprehension.

Reply:
Addressed by https://github.com/digital-asset/daml/pull/2713/commits/f62f594d2f3bc634db569bed7b35d483eb772153

Thanks for catching this, I have no idea of why I used for-comprehensions here. I guess I had slightly more complex stuff that I then deleted.

Reply:
Addressed by https://github.com/digital-asset/daml/pull/2713/commits/f62f594d2f3bc634db569bed7b35d483eb772153

Thanks for catching this, my best guess is I left those over from a previous iteration.